test/object-store/util/test_utils.hpp

```diff
@@ -58,11 +58,14 @@ class TestingStateMachine {
     void transition_with(Func&& func)
     {
         {
-            std::lock_guard lock{m_mutex};
-            std::optional<E> new_state = func(m_cur_state);
+            std::unique_lock lock{m_mutex};
+            E cur_state = m_cur_state;
+            lock.unlock();
+            std::optional<E> new_state = func(cur_state);
```

Comment:
The point of this function was to guarantee that you were updating the state of the state machine while holding the mutex so nobody else could change the state while you were evaluating whether to move to a new state. So this change indicates to me that there's something broken in one of your tests. What kind of deadlock were you hitting that led you to making this change?

Reply:
It was a TSAN failure around checking the base URL inside the `transition_with()` func handler. I reverted this code and updated the test.